cmd/testbeacon.go

```diff
@@ -0,0 +1,124 @@
+// Copyright © 2022-2024 Obol Labs Inc. Licensed under the terms of a Business Source License 1.1
+
+package cmd
+
+import (
+	"context"
+	"io"
+	"sort"
+	"time"
+
+	"github.com/spf13/cobra"
+	"golang.org/x/exp/maps"
+
+	"github.com/obolnetwork/charon/app/errors"
+)
+
+type testBeaconConfig struct {
+	testConfig
+	Endpoints []string
+}
+
+func newTestBeaconCmd(runFunc func(context.Context, io.Writer, testBeaconConfig) error) *cobra.Command {
+	var config testBeaconConfig
+
+	cmd := &cobra.Command{
+		Use:   "beacon",
+		Short: "Run multiple tests towards beacon nodes",
+		Long:  `Run multiple tests towards beacon nodes. Verify if the current setup is suitable for mainnet cluster.`,
+		Args:  cobra.NoArgs,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return runFunc(cmd.Context(), cmd.OutOrStdout(), config)
+		},
+	}
+
+	bindTestFlags(cmd, &config.testConfig)
+	bindTestBeaconFlags(cmd, &config)
+
+	return cmd
+}
+
+func bindTestBeaconFlags(cmd *cobra.Command, config *testBeaconConfig) {
+	const endpoints = "endpoints"
+	cmd.Flags().StringSliceVar(&config.Endpoints, endpoints, nil, "[REQUIRED] Comma separated list of one or more beacon node endpoint URLs.")
+	mustMarkFlagRequired(cmd, endpoints)
+}
+
+func supportedBeaconTestCases() map[testCaseName]func(*testBeaconConfig) testResult {
+	return map[testCaseName]func(*testBeaconConfig) testResult{
+		{name: "ping", order: 1}: beaconPing,
+	}
+}
+
+func runTestBeacon(ctx context.Context, w io.Writer, cfg testBeaconConfig) (err error) {
+	testCases := supportedBeaconTestCases()
+	supportedTestCases := maps.Keys(testCases)
+
+	queuedTests, err := filterTests(supportedTestCases, cfg.testConfig)
+	if err != nil {
+		return err
+	}
+	sort.Slice(queuedTests, func(i, j int) bool {
+		return queuedTests[i].order < queuedTests[j].order
+	})
+
+	parentCtx := ctx
+	if parentCtx == nil {
```

Comment:
Parent `ctx` must never be `nil` if we follow best practices. At least for un-exported functions.

Reply:
This parent context is coming from cobra's `cmd.Context()`, which says:

```
// Context returns underlying command context. If command was executed
// with ExecuteContext or the context was set with SetContext, the
// previously set context will be returned. Otherwise, nil is returned.
```

Reply:
I believe we shall always use ExecuteContext, especially for tests, to support graceful shutdown (catching SIGTERM, etc and canceling the root context).

See the skeleton here: https://github.com/pinebit/go-graceful-shutdown

Reply:
Yes, makes sense to me. Looking at it now though, it will require quite a bit of refactoring on the main CLI. Do you think it's okay to leave it like that and open an issue to change it everywhere where suitable (not only test)? Or shall I do this refactoring in this PR?

Reply:
Let's have a separate ticket to ensure we implemented graceful shutdown for charon. I did not check, maybe we have it but not for all commands?
For these tests we must have it 100%, because for continuous tests the user will terminate the app with docker down or ctrl+c/z.

Reply:
This will guarantee the passed context is not `nil`. But until we have this done, leave your check and fallback to Background().

Reply:
Agreed with the proposed solution (re-checking for context usage and nilness in CLI).

>Let's have a separate ticket to ensure we implemented graceful shutdown for charon

Charon has a lifecycle manager in `run` but not in CLI commands, which are supposed to run as one-shot instances rather than long-term.